File: android/src/main/java/com/sendbird/calls/reactnative/RNSBDirectCallVideoView.kt
package com.sendbird.calls.reactnative

import android.content.Context
import com.sendbird.calls.DirectCall
import com.sendbird.calls.SendBirdCall
import com.sendbird.calls.reactnative.view.BaseVideoView

enum class ViewType {
    LOCAL, REMOTE
}

class RNSBDirectCallVideoView(context: Context) : BaseVideoView(context) {
    private var mViewType = ViewType.LOCAL
    private var mCallId: String? = null

    private fun getCall(id: String?): DirectCall? {
        val callId = id ?: mCallId ?: return null
        return SendBirdCall.getCall(callId)
    }

    fun updateView() {
        val call = getCall(mCallId)
        when (mViewType) {
            ViewType.REMOTE -> call?.setRemoteVideoView(mSurface)
            ViewType.LOCAL -> call?.setLocalVideoView(mSurface)
        }
    }

    fun setZOrderMediaOverlay(overlay: Boolean) {
        mSurface.setZOrderMediaOverlay(overlay)
    }

    fun setViewType(viewType: ViewType) {
        this.mViewType = viewType
        this.updateView()
    }

    fun setCallId(callId: String) {
        this.mCallId = callId
        this.updateView()
    }
}

File: ios/RNSBDirectCallVideoView.swift
//
//  RNSBDirectCallVideoView.swift
//  RNSendbirdCalls
//
//  Created by Airen Kang on 2022/06/11.
//  Copyright © 2022 Sendbird. All rights reserved.
//

import Foundation
import React
import SendBirdCalls

enum ViewType: String {
    case remote
    case local
}

@objc(RNSBDirectCallVideoView)
class RNSBDirectCallVideoView: BaseVideoView {
    @objc var callId: String? = nil
    @objc var viewType: String = ViewType.local.rawValue
    
    private func getCall(id: String?) -> DirectCall? {
        guard let callId = callId else { return nil }
        
        if let call = SendBirdCall.getCall(forCallId: callId) {
            return call
        }
        
        if let uuid = UUID(uuidString: callId), let call = SendBirdCall.getCall(forUUID: uuid) {
            return call
        }
        
        return nil
    }
    
    
    override func didSetProps(_ changedProps: [String]!) {
        if changedProps.contains("viewType") {
            setViewType()
        }
        
        if changedProps.contains("callId") {
            setCallId()
        }
    }
    
    private func setViewType() {
        guard let call = getCall(id: self.callId),
              let viewType = ViewType(rawValue: self.viewType.lowercased()) else { return }
        
        switch(viewType){
        case .remote:
            call.updateRemoteVideoView(self.surface)
        case .local:
            call.updateLocalVideoView(self.surface)
        }
    }
    
    private func setCallId() {
        guard let call = getCall(id: self.callId),
              let viewType = ViewType(rawValue: self.viewType.lowercased()) else { return }
        
        switch(viewType){
        case .remote:
            call.updateRemoteVideoView(self.surface)
        case .local:
            call.updateLocalVideoView(self.surface)
        }
    }
    
}
